### PR TITLE
ETag demo in logging sample

### PR DESCRIPTION
**Pull request by achamayou**
Demo illustrating the use of ETags in a CCF application, #6081, see [end to end test for usage sample](https://github.com/microsoft/CCF/pull/6110/files#diff-3b229b160527d90c259f61604ec7cd0d96f94449a5131c4c8eac0825d2e03b13L1800).

Add request headers `If-Match` and `If-None-Match`, as well as response header `ETag` on:

- `POST /app/log/public`
- `GET /app/log/public?id={id}`
- `DELETE /app/log/public?id={id}`

Documented as part of the C++ app sample.

### CODE REVIEW

File: samples/apps/logging/logging.cpp [hunk 1]
```diff
@@ -11,13 +11,15 @@
...
 #include "ccf/historical_queries_adapter.h"
+#include "ccf/http_etag.h"
 #include "ccf/http_query.h"
 #include "ccf/indexing/strategies/seqnos_by_key_bucketed.h"
 #include "ccf/indexing/strategy.h"
 #include "ccf/json_handler.h"
 #include "ccf/version.h"
 
 #include <charconv>
+#include <regex>
```
Comment (eddyashton) [inline samples/apps/logging/logging.cpp:22]:
```suggestion
```

Think we can remove this here, as it's only used in `http_etag.h`

File: tests/e2e_logging.py [hunk 2]
```diff
@@ -1798,6 +1798,168 @@ def test_basic_constraints(network, args):
...
+    primary, _ = network.find_primary()
+
+    with primary.client("user0") as c:
+        doc = {"id": 999999, "msg": "hello world"}
+        etag = sha256(doc["msg"].encode()).hexdigest()
+
+        # POST ETag matches value
+        r = c.post("/app/log/public", doc)
+        assert r.status_code == http.HTTPStatus.OK
+        assert r.headers["ETag"] == etag, r.headers["ETag"]
```
Comment (eddyashton) [inline tests/e2e_logging.py:1811]:
```suggestion

        # POST ETag matches value
        r = c.post("/app/log/public", doc)
        assert r.status_code == http.HTTPStatus.OK
        assert "ETag" in r.headers, r.headers
        etag = r.headers["ETag"]
```

I don't know if should make any assertions about the ETag's value? We happen to know how it's constructed in this instance, but in general its an opaque identifier.

File: tests/e2e_logging.py [hunk 3]
```diff
@@ -1798,6 +1798,168 @@ def test_basic_constraints(network, args):
...
+        assert r.status_code == http.HTTPStatus.PRECONDITION_FAILED
+
+        # POST If-Match: matching ETag returns 200
+        r = c.post("/app/log/public", doc, headers={"If-Match": f'"{etag}"'})
+        assert r.status_code == http.HTTPStatus.OK
+        etag = sha256(doc["msg"].encode()).hexdigest()
+        assert r.headers["ETag"] == etag, r.headers["ETag"]
+
+        # POST If-Match: mutiple, one which is matching, ETags returns 200
+        r = c.post("/app/log/public", doc, headers={"If-Match": f'"{etag}", "abc"'})
```
Comment (eddyashton) [inline tests/e2e_logging.py:1857]: Other test cases we could add to be thorough:
```
If-Match: "abc", <ETag>  # ETag matches, but is not first in list of multiple
If-Match: "abc", "def"      # Multiple ETags, none match
```

File: js/ccf-app/src/crypto.ts [hunk 4]
```diff
@@ -108,7 +108,7 @@ export const eddsaPemToJwk = ccf.crypto.eddsaPemToJwk;
 export const pubJwkToPem = ccf.crypto.pubJwkToPem;
 
 /**
- * @inheritDoc global!CCFCrypto.JwkToPem
+ * @inheritDoc global!CCFCrypto.jwkToPem
```
Comment (eddyashton) [inline js/ccf-app/src/crypto.ts:111]: These docs fixes are good, but can we put them in a separate PR?

File: include/ccf/http_etag.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          return;
+        }
+
+        if (match_header.value() == "*")
+        {
+          any_value = true;
+          return;
+        }
+
+        std::regex etag_rx("\\\"([0-9a-f]+)\\\",?\\s*");
```
Comment (eddyashton) [inline include/ccf/http_etag.h:46]: Trying to understand if this regex is necessary, or overkill. Could we get the same result with:
```
etag_values = nonstd::split(match_header.value(), ",")
foreach tag_value in etag_values:
  Strip whitespace
  Check it starts and ends with "
  Append to if_etags
```

This would be more readable than a (heavily-escaped) regex, if I'm not missing anything.

File: include/ccf/http_etag.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      bool any_value = false;
+      /// If-Match header is present and has specific etag values
+      std::set<std::string> if_etags;
+
+    public:
+      /** Construct a Matcher from a match header if present
+       *
+       * Note: Weak tags are not supported.
+       */
+      Matcher(const std::optional<std::string>& match_header)
```
Comment (eddyashton) [inline include/ccf/http_etag.h:32]: Could this take a `string` rather than an `optional`, _only_ be constructed when the header was present, and then this matcher could drop all of its `empty` state/code? I think all of the paths are supported identical (checks on `.empty()` on a `optional<string>` rather than a `Matcher`, an error if both _headers_ are present, errors thrown if the content is not a valid ETags list), just in maybe different orders.

File: tests/e2e_logging.py (hunk 2 above)
Comment (achamayou) [reply tests/e2e_logging.py:1811]: I general sure, in this particular test, it seems reasonable to be precise?

File: js/ccf-app/src/crypto.ts (hunk 4 above)
Comment (achamayou) [reply js/ccf-app/src/crypto.ts:111]: Sure, they were just drive by fixing of Sphinx warnings.

File: include/ccf/http_etag.h (hunk 5 above)
Comment (achamayou) [reply include/ccf/http_etag.h:46]: We probably could, I started down this road because I was hoping to capture all the values in one match(), but the API is awful and that's just not possible. The rest is path dependent.

File: include/ccf/http_etag.h (hunk 6 above)
Comment (achamayou) [reply include/ccf/http_etag.h:32]: Probably, I started that way but this seemed to reduce the number of branches somewhat. I'll have another look.

File: js/ccf-app/src/crypto.ts (hunk 4 above)
Comment (achamayou) [reply js/ccf-app/src/crypto.ts:111]: #6120
